### PR TITLE
LastFM Redo

### PR DESCRIPTION
**Pull request by coredev-uk**

### CODE REVIEW

**Comment by coredev-uk** [comment]
If any of the dev team could test this it would be great.

**Comment by cryptofyre** [comment]
I'll check it out.

**Comment by coredev-uk** [comment]
> I'll check it out.

Thanks. The only thing to note currently is incognito/private mode.

**Comment by GamingLiamStudios** [comment]
Not very extensive testing, but half an hour of songs later & it doesn't look like it's stopped working properly. All songs were scrobbled apart from the first one right after I reauthed.

**Comment by coredev-uk** [comment]
> Not very extensive testing, but half an hour of songs later & it doesn't look like it's stopped working properly. All songs were scrobbled apart from the first one right after I reauthed.

Yeah I've noticed the reauth thing, usually it only gets scrobbled/initiated through now playing when the song starts with the nowPlayingDidChange. I could put a workaround in for it, but it would be a bit unnecessary for one song.

**Comment by GamingLiamStudios** [comment]
Looks like Looped Scrobbles are filtered by default. Is there any way of disabling this?

**Comment by coredev-uk** [comment]
Yeah should be easy to re-add the config option back.

**Review by cryptofyre — APPROVED** [review]
Tested for a day, runs good.

**Comment by GamingLiamStudios** [comment]
I'm not sure the `filter loop` button works... Been playing the same song on loop for the last 20-ish minutes with the option both on and off with no change over on last.fm

**Comment by coredev-uk** [comment]
> I'm not sure the `filter loop` button works... Been playing the same song on loop for the last 20-ish minutes with the option both on and off with no change over on last.fm

After just testing it myself, its working correctly. Not sure why you are experiencing issues.

**Review by maikirakiwi — APPROVED** [review]

**Comment by coredev-uk** [comment]
Pending tasks before merge:

- [ ] Further testing on filter loop and connection
- [ ] Configuration option to prevent scrobbling of certain media types.

File: resources/lastfm.ts [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+                this._lastfm.cachedNowPlayingAttributes = attributes
+            }
+
+        } else {
+            this.authenticate()
+        }
+    }
+
+    private getAlbumName(attributes: any): string {
+        return attributes.albumName.replace(/ - Single| - EP/g, '');
```
Comment (UnbreakCode) [inline resources/lastfm.ts:162]: Maybe also replace (Official Video)

**Comment by coredev-uk** [comment]
Media types filter done. If anyone has more types to add, please say. Some further testing needed before merge.

And @booploops if you could improve upon the selection interface for filter media types in the renderer at some point, that would be great.

File: src/main/base/store.ts [hunk 2]
```diff
@@ -323,43 +359,4 @@ export class Store {
...
-            })
-        }
-        let postBody = {
-            id: Store.cfg.get('connectUser.id'),
-            app: electron.app.getName(),
-            version: electron.app.isPackaged ? electron.app.getVersion() : 'dev',
-            syncData: syncData
-        }
-
-        fetch('https://connect.cidercollective.dev/api/v1/setttings/set', {
```
Comment (maikirakiwi) [inline src/main/base/store.ts:357]: "settings" with 3 t's

Comment (coredev-uk) [reply src/main/base/store.ts:357]: Thats not my doing, and the api page setup for it also has 3 t's... @quacksire

**Comment by GamingLiamStudios** [comment]
Tested the loops again, works mostly fine now. Only song it doesn't work with(or even scribble) is 13 by Tally Hall, a 13-second long song.

**Comment by coredev-uk** [comment]
> Tested the loops again, works mostly fine now. Only song it didn't work with(or even scrobble) is 13 by Tally Hall, a 13-second long song.

The song is probably too short for the lastfm api to accept.

**Comment by GamingLiamStudios** [comment]
oh yeah and I have to manually remove the `connectivity` section of `cider-config.json` if I want to use the `main` branch

**Comment by GamingLiamStudios** [comment]
> > Tested the loops again, works mostly fine now. Only song it didn't work with(or even scrobble) is 13 by Tally Hall, a 13-second long song.
> 
> The song is probably too short for the lastfm api to accept.

Works fine on the `main` branch

**Comment by coredev-uk** [comment]
Interesting, are any errors being displayed?

**Comment by GamingLiamStudios** [comment]
> Interesting, are any errors being displayed?

for?

**Comment by coredev-uk** [comment]
You sure its not working, I've just managed to get it to scrobble and update now playing. 
image

**Comment by GamingLiamStudios** [comment]
Last.fm is recognising that I am scrobbling the song, but it's not saving it.
image
Other songs(over 30s) seem to work fine. The problem has something to do with the new lastfm code too, since it scrobbled & saves just fine on the `main` branch

**Comment by coredev-uk** [comment]
> Last.fm is recognising that I am scrobbling the song, but it's not saving it. image Other songs(over 30s) seem to work fine. The problem has something to do with the new lastfm code too, since it scrobbled & saves just fine on the `main` branch

Make sure you're running it in an uncompiled state if you want to get more logging, the console.debug function only runs in that state.

**Comment by Monochromish** [comment]
Don't know if it's the right place to ask this, but is it possible to provide a token query in the last.fm connect application link as well? Since without it, I am not able to connect Cider to scrobble for the past few weeks.
The `Yes, Allow Access` button on last.fm Connect Application page does nothing besides redirect to Cider unless I provide a token query. Even DeadFrost was having the same problem.

**Comment by coredev-uk** [comment]
> Don't know if it's the right place to ask this, but is it possible to provide a token query in the last.fm connect application link as well? Since without it, I am not able to connect Cider to scrobble for the past few weeks. The `Yes, Allow Access` button on last.fm Connect Application page does nothing besides redirect to Cider unless I provide a token query. Even DeadFrost was having the same problem.

Usually the protocol does work. I have implemented a manual field that allows you to input the token. You just have to remove the callback from the api url.

**Comment by coredev-uk** [comment]
Merging this now, I'm all done with my tests.

**Pull request merged**